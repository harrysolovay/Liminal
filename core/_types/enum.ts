import type { Type } from "../Type.ts"
import { declareType } from "./_declareType.ts"

function enum_<V extends Array<string>>(...values: V): Type<V[number]> {
  return declareType(enum_<V>, values.toSorted())
}
Object.defineProperty(enum_, "name", { value: "enum" })
export { enum_ as enum }
