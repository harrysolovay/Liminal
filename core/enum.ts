import type { Type } from "../Type.ts"
import { declare } from "../TypeDeclaration.ts"

export { enum_ as enum }
export function enum_<K extends string>(...members: Array<K>): Type.Initial<K> {
  return declare({
    name: "enum",
    source: {
      factory: enum_,
      args: { members },
    },
    subschema: () => ({
      type: "string",
      enum: members,
    }),
  })
}
Object.defineProperty(enum_, "name", { value: "enum" })
