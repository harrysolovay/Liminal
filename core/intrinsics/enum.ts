import { assert, assertArrayIncludes } from "@std/assert"
import { declare } from "../declare.ts"
import type { Type } from "../Type.ts"

export { enum_ as enum }
function enum_<V extends Array<string>>(...values: V): Type<V[number], never> {
  return declare({
    factory: enum_,
    args: values.toSorted(),
    assert: (value) => {
      assert(typeof value === "string")
      assertArrayIncludes(values, [value])
    },
  })
}
Object.defineProperty(enum_, "name", { value: "enum" })
