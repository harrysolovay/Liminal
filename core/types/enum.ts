import { Type } from "../Type.ts"

export { enum_ as enum }
function enum_<K extends string>(...values: Array<K>): Type<K> {
  return Type({
    kind: "enum",
    factory: enum_,
    args: values,
    argsLookup: { values },
  })
}
Object.defineProperty(enum_, "name", { value: "enum" })
