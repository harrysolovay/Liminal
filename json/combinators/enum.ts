import { Type } from "../../core/mod.ts"

export { enum_ as enum }
function enum_<K extends string>(...members: Array<K>): Type<K> {
  return Type({
    factory: enum_,
    args: members,
  })
}
Object.defineProperty(enum_, "name", { value: "enum" })
