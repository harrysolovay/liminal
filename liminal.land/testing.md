# Liminal Testing <Badge type="warning" text="beta" />
