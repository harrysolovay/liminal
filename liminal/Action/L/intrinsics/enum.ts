import type { L } from "../L.js"
import { declare } from "../declare.js"

// `union` instead?
function enum_<V extends Array<string>>(...values: V): L<V[number]> {
  return declare(enum_<V>, values.toSorted())
}
Object.defineProperty(enum_, "name", { value: "enum" })
export { enum_ as enum }
