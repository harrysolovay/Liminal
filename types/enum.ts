import { declareType, type Type } from "../core/mod.ts"

export { enum_ as enum }
function enum_<K extends string>(...members: Array<K>): Type<K> {
  return declareType({
    name: "enum",
    source: {
      factory: enum_,
      args: members,
    },
  })
}
Object.defineProperty(enum_, "name", { value: "enum" })
