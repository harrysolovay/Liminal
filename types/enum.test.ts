import { assertTypeSnapshot } from "testing"
import { enum as enum_ } from "./enum.ts"

Deno.test("enum", async (t) => {
  await assertTypeSnapshot(t, enum_("A", "B", "C"))
})
