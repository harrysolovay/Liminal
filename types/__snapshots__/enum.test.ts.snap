export const snapshot = {};

snapshot[`enum 1`] = `
{
  enum: [
    "A",
    "B",
    "C",
  ],
  type: "string",
}
`;

snapshot[`enum 2`] = `
{
  description: "One.",
  enum: [
    "A",
    "B",
    "C",
  ],
  type: "string",
}
`;

snapshot[`enum 3`] = `
{
  description: "One. Two.",
  enum: [
    "A",
    "B",
    "C",
  ],
  type: "string",
}
`;
